pento/free-as-in-speech
# Enable development without running the browser extension

This Pull Request adds the ability to use [HAR files](https://en.wikipedia.org/wiki/HAR_(file_format)) for local development.

By recording an HAR file using a browser's developer tools (usually there is something like "Save All as HAR" that can be used for that), either after having visited a supported site, or after an export took place, this HAR file can then be used for continued, local (=offline) development.

Specifically, this adds a CLI script `export-from-har.js` that accepts a HAR file as argument and then runs the business logic of exporting the site. Contrary to running in the browser and thus making HTTP requests to the real URLs, the requests are satisfied (if possible) by using the specified HAR file.

This also allows quick testing with different sites by swapping out the locally saved HAR files.

To achieve this workflow, we have created a package `fetch-from-har` that implements a polyfill for `fetch()` (which is not available in node anyway and would need its own polyfill) which is initialized with the HAR file to load the data from.

### Testing Instructions

In order to bridge the compatibility gap between ES6 used in the extension, and common JS used in node scripts, we're using webpack, hence during development you need to keep a `npm run watch` running in a tab.

The script can then be executed with either of these:
```
// extract media manager files and the communities-blog-app:
$ npm run extract-wix my.har -- -a 14bcded7-0066-7c35-14d7-466cb3f09103,media-manager
// extract media manager files and the communities-blog-app:
$ npm run extract-wix my.har -- -a media-manager
// if you call the node file directly, you don't have to do the -- trick
$ node build/export-from-har.js wix my.har -a 14bcded7-0066-7c35-14d7-466cb3f09103
```

For the moment, it'll output the exported WXR to the console.

As a side-product of the first implementation using `fetch-mock` (see #17), this also adds a jest test case.
